File: test_acJobsApp.py
"""test_acJobsApp.py - Esempio d'uso della libreria acJobsApp (Python).

Uso:
    python test_acJobsApp.py [percorso_ini]

Usa il file condiviso test_acJobsApp.ini (NON lo crea, NON lo cancella):
esegue i job con Start -> Run -> End e pulisce solo .end e .log generati.
La callback supporta COMMAND=SHELL (esegue PARAM.CMD) e, in alternativa,
esegue direttamente il valore di COMMAND come comando di shell
(es. COMMAND=cmd.exe /c dir *.* /b).
"""
import os
import subprocess
import sys

from acJobsApp import acJobsApp

MAX_OUT = 2000


def run_shell(sCmd, sWorkDir):
    """Esegue un comando di shell, ritorna (errore, output)."""
    try:
        p = subprocess.run(sCmd, shell=True, capture_output=True, text=True,
                           cwd=sWorkDir, timeout=120, errors="replace")
        sOut = (p.stdout or "") + (p.stderr or "")
        if len(sOut) > MAX_OUT:
            sOut = sOut[:MAX_OUT]
        if p.returncode != 0:
            return f"exit={p.returncode}", sOut
        return "", sOut
    except Exception as e:
        return f"Errore shell: {e}", ""


def main():
    sBase = os.path.dirname(os.path.abspath(__file__))
    if len(sys.argv) < 2:
        sys.argv.append(os.path.join(sBase, "test_acJobsApp.ini"))
    sIni = sys.argv[1]
    if not os.path.isfile(sIni):
        print(f"File di prova non trovato: {sIni}")
        return

    jData = acJobsApp()

    def cbCommands(dJob):
        sCmd = dJob.get("COMMAND", "")
        if sCmd == "SHELL":
            sShell = dJob.get("PARAM.CMD", "")
            if not sShell:
                return jData.Return("PARAM.CMD mancante", "")
            sErr, sOut = run_shell(sShell, sBase)
            return jData.Return(sErr, sOut)
        if sCmd:
            sErr, sOut = run_shell(sCmd, sBase)
            return jData.Return(sErr, sOut)
        return jData.Return("COMMAND mancante", "")

    sResult = jData.Start()
    # Nota: End() chiama sys.exit() in caso di errore: intercetta SystemExit
    # cosi' la pulizia degli artefatti viene eseguita comunque.
    nCode = 0
    try:
        if sResult != "":
            nCode = jData.End(sResult)
        else:
            nCode = jData.End(jData.Run(cbCommands))
    except SystemExit as e:
        nCode = e.code if isinstance(e.code, int) else 2
    if nCode is None:
        nCode = 0  # End() ritorna None su successo (exit 0 senza sys.exit)
    print(f"Exit code: {nCode}")

    # Pulizia dei soli artefatti generati (l'ini condiviso resta)
    base, _ = os.path.splitext(sIni)
    for f in (base + ".end", os.path.join(sBase, "test_acJobsApp.log")):
        try:
            if os.path.isfile(f):
                os.remove(f)
        except OSError:
            pass
    print("Test completato.")


if __name__ == "__main__":
    main()

File: test_acJobsApp.java
// test_acJobsApp.java - Esempio d'uso della libreria acJobsApp.java
// Compilazione ed esecuzione (serve un JDK):
//   javac acJobsApp.java test_acJobsApp.java
//   java test_acJobsApp
// L'esempio: crea un file .ini di prova, esegue 2 job (SALUTA e SOMMA),
// scrive il file .end, mostra un poll execReturn con timeout breve e ripulisce.

import java.io.File;
import java.util.LinkedHashMap;
import java.util.Map;

public class test_acJobsApp {

    public static void main(String[] args) {
        // --- 1. Funzioni di supporto ---
        System.out.println("Timestamp : " + acJobsApp.timestamp());
        System.out.println("Expand    : " + acJobsApp.expand("Ciao $USER da $SYS.OS",
                new LinkedHashMap<String, String>() {{ put("USER", "Mario"); }}));
        System.out.println("Bool TRUE : " + acJobsApp.stringBool("TRUE"));

        // --- 2. Crea un file .ini di prova (solo PARAM.*, nessun FILE.* richiesto) ---
        String sIni = acJobsApp.normalizePath(new File(System.getProperty("user.dir", "."), "test_lavoro.ini").getPath());
        Map<String, Map<String, String>> data = new LinkedHashMap<String, Map<String, String>>();
        Map<String, String> cfg = new LinkedHashMap<String, String>();
        cfg.put("TYPE", "NTJOBS.APP.1.0");
        cfg.put("NAME", "DEMO_JAVA");
        cfg.put("EXIT", "TRUE");
        cfg.put("LOG", "test_demo.log");
        cfg.put("BASE_DIR", "C:\\dati");
        data.put("CONFIG", cfg);
        Map<String, String> job1 = new LinkedHashMap<String, String>();
        job1.put("COMMAND", "SALUTA");
        job1.put("PARAM.NAME", "Mario");
        job1.put("PARAM.DIR", "$BASE_DIR");
        data.put("JOB1", job1);
        Map<String, String> job2 = new LinkedHashMap<String, String>();
        job2.put("COMMAND", "SOMMA");
        job2.put("PARAM.A", "40");
        job2.put("PARAM.B", "2");
        data.put("JOB2", job2);
        String sErr = acJobsApp.saveDictToIni(data, sIni);
        if (sErr != null && !sErr.isEmpty()) { System.out.println("Errore scrittura ini: " + sErr); return; }

        // --- 3. Ciclo di vita: Start -> Run -> End ---
        acJobsApp jData = new acJobsApp();
        String sResult = jData.start(new String[]{sIni});
        int nExit;
        if (sResult != null && !sResult.isEmpty()) {
            nExit = jData.end(sResult);
        } else {
            nExit = jData.end(jData.run(dJob -> {
                String sCmd = dJob.containsKey("COMMAND") ? dJob.get("COMMAND") : "";
                if ("SALUTA".equals(sCmd)) {
                    String sNome = dJob.containsKey("PARAM.NAME") ? dJob.get("PARAM.NAME") : "mondo";
                    return jData.jobReturn("", "Ciao " + sNome + "!");
                }
                if ("SOMMA".equals(sCmd)) {
                    try {
                        int a = Integer.parseInt(dJob.get("PARAM.A").trim());
                        int b = Integer.parseInt(dJob.get("PARAM.B").trim());
                        return jData.jobReturn("", "Somma=" + (a + b));
                    } catch (Exception e) {
                        return jData.jobReturn("Parametri non numerici", "");
                    }
                }
                return jData.jobReturn("Comando sconosciuto: " + sCmd, "");
            }));
        }
        System.out.println("Exit code: " + nExit);

        // --- 4. Leggi il file .end e mostra l'esito globale ---
        String sEnd = sIni.substring(0, sIni.lastIndexOf('.')) + ".end";
        acJobsApp.IniRead rd = acJobsApp.readIniToDict(sEnd);
        if (rd.err == null || rd.err.isEmpty()) {
            Map<String, String> endCfg = rd.data.get("CONFIG");
            System.out.println("END CONFIG RETURN.TYPE  = " + (endCfg != null ? endCfg.get("RETURN.TYPE") : "?"));
            System.out.println("END JOB1 RETURN.VALUE   = " + rd.data.get("JOB1").get("RETURN.VALUE"));
            System.out.println("END JOB2 RETURN.VALUE   = " + rd.data.get("JOB2").get("RETURN.VALUE"));
        }

        // --- 5. execReturn con timeout breve su ID inesistente -> mappa vuota (non finito) ---
        Map<String, Map<String, String>> pending = jData.execReturn("id_inesistente_xyz", 1);
        System.out.println("execReturn su ID inesistente, chiavi=" + pending.size() + " (0 = non finito, atteso)");

        // Esempio di lancio reale (decommentare con uno script esistente):
        // String sLaunch = jData.exec("C:/apps/figlia.py",
        //         new LinkedHashMap<String, String>() {{ put("NAME", "FIGLIA"); }},
        //         new LinkedHashMap<String, Object>() {{ put("JOB1", new LinkedHashMap<String, String>() {{ put("COMMAND", "SALUTA"); }}); }},
        //         "lotto1");
        // Map<String, Map<String, String>> dictResult = jData.execReturn("lotto1", 30);

        // --- 6. Pulizia file di prova ---
        new File(sIni).delete();
        new File(sEnd).delete();
        new File("test_demo.log").delete();
        System.out.println("Test completato.");
    }
}
